enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,1156 @@
+             * Backup `backups/current-commit.id/`, continue running off `microshift/`
+           - `version.json` does not match `current-commit.id`
+             > Migrated without backup? Bug or user interference.
+             * Exit 1, block `microshift run`
+
+       - `current-commit.system == unhealthy`
+         > `prev-boot-commit` was staged over `unhealthy` `current-commit` without metadata cleanup.
+         > We do no support upgrade from unhealthy commit.
+         > Admin should've heal previous commit or delete MicroShift data before upgrade to start from clean state.
+         >
+         > This should be caught by *`prev-boot-commit.system` is `unhealthy` > `current-commit` != `prev-boot-commit` >
+         > `current-commit` does not exist in `health.json`*
+         * Exit 1, block `microshift run`
+
+   - `current-commit == prev-boot-commit`
+     > Because `current-commit` is already present in `health.json`, it's 2nd, 3rd,... boot of the commit (not 1st).
+
+     - `backups/current-commit.id/` exists
+       > commit was healthy at least once already.
+       > It means greenboot removed `boot_counter` and current boot isn't due to greenboot's automatic reboot.
+       > After addressing issues, admin should mark commit as healthy, so this shouldn't happen.
+       > Let's assume admin addressed the issues but forgot to mark as healthy and just rebooted the device.
+       * No backup, no restore, skip migration, allow `microshift run`
+         > If admin addressed issues, it should be healthy again.
+         > If issue persists, it'll end up unhealthy requiring manual intervention
+         > (which is the same result if we'd block `microshift run`).
+
+     - `backups/current-commit.id/` doesn't exist
+       > commit was always unhealthy - it wasn't "accepted" by greenboot, so it might still rollback.
+       > Because admin should heal the commit before manually rebooting, we assume it's reboot due to greenboot.
+
+       - `earlier-commit not in health.json`
+         > For MicroShift it's first commit it's running, but it doesn't mean it's the only commit
+         > (rollback commit might be without MicroShift).
+         >
+         > ostree status:
+         > 1. `current-commit` == `prev-boot-commit` - currently running,
+         >    last boot unhealthy, greenboot didn't accept it yet so it might still rollback
+         > 1. (optionally) rollback (possibly without MicroShift)
+         * Delete data, try from clean state
+
+       - `earlier-commit in health.json`
+         > For MicroShift it looks like it ran already on the device, but `earlier-commit` might or might not be
+         > the rollback commit.
+         >
+         > ostree status:
+         > 1. `current-commit` == `prev-boot-commit` - currently running, last boot unhealthy, greenboot didn't accept it yet
+         > 1. (auto) rollback (might be `earlier` or not)
+
+         - `earlier-commit == ostree rollback commit`
```

Comment:
I haven't found it in output of `rpm-ostree status --json` (maybe it's implicit based on some fields), but `ostree admin status` reports `(rollback)` so it can be found
```
$ ostree admin status
* fedora 086fccc8c374b81040270ee8a2b0f05b31e496da1f033daf3fc97263cbd5daf8.0
    Version: 38.20230509.0
    origin: <unknown origin type>
  fedora f066091d0e7e11f96465b512115ba11684904efc6a9e4d42c877daec6c8ec591.0 (rollback)
    Version: 38.20230509.0
    origin: <unknown origin type>
  fedora fa62e262bd5d8afe7de1fda8f2865cee3a1734c62d299bab4e2e1d00f4c33f3e.0
    Version: 38.20230509.0
    Pinned: yes
    origin refspec: fedora:fedora/38/x86_64/silverblue
    GPG: Signature made Tue 09 May 2023 02:47:11 AM CEST using RSA key ID 809A8D7CEB10B464
    GPG: Good signature from "Fedora <fedora-38-primary@fedoraproject.org>"
```

Here's how ostree finds `pending` and `rollback` deployments https://github.com/ostreedev/ostree/blob/main/src/libostree/ostree-sysroot.c#L1554
Looks like it's matter of ordering: deployment before booted is pending, deploy after booted is rollback

After rolling back:
```
* fedora f066091d0e7e11f96465b512115ba11684904efc6a9e4d42c877daec6c8ec591.0
    Version: 38.20230509.0
    origin: <unknown origin type>
  fedora 086fccc8c374b81040270ee8a2b0f05b31e496da1f033daf3fc97263cbd5daf8.0 (rollback)
    Version: 38.20230509.0
    origin: <unknown origin type>
  fedora fa62e262bd5d8afe7de1fda8f2865cee3a1734c62d299bab4e2e1d00f4c33f3e.0
    Version: 38.20230509.0
    Pinned: yes
    origin refspec: fedora:fedora/38/x86_64/silverblue
    GPG: Signature made Tue 09 May 2023 02:47:11 AM CEST using RSA key ID 809A8D7CEB10B464
    GPG: Good signature from "Fedora <fedora-38-primary@fedoraproject.org>"
```

There might be more caveats to (rpm-)ostree when it comes to how deployments are staged etc